packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp

```diff
@@ -0,0 +1,499 @@
+      if(regQueue.empty()) curr = &artQueue;
+      else curr = &regQueue;
+      while(!curr->empty()){
+	label_t curr_node = femvData[curr->front()];
+        curr->pop();
+        
+        //if the current node is a copy, it shouldn't propagate out to its neighbors.
+        if(curr_node.id >= nLocalOwned) continue;
+
+        lno_t out_degree = g->out_degree(curr_node.id);
+        lno_t* outs = g->out_vertices(curr_node.id);
+        for(int i = 0; i < out_degree; i++){
+	  label_t neighbor = femvData[outs[i]];
+	  IcePropGrounding_Status old_gs = neighbor.getGroundingStatus();
+          
+          //give curr_node's neighbor some more labels
+          giveLabels(curr_node, neighbor);
+          
+          if(old_gs != neighbor.getGroundingStatus()){
+            femv->replaceLocalValue(outs[i],0,neighbor);
+            if(neighbor.is_art) artQueue.push(neighbor.id);
+            else regQueue.push(neighbor.id);
+          }
+        }
+        if(curr->empty()){
+          if(curr == &regQueue) curr = &artQueue;
+          else curr = &regQueue;
+        }
+        
+      }
+      
+      //store old vertex statuses for detecting changes
+      std::vector<IcePropGrounding_Status> old_statuses(nLocalOwned,ICEPROPGS_NONE);      
+      for(size_t i = 0; i < nLocalOwned; i ++){
+        old_statuses[i] = femvData[i].getGroundingStatus();
+      }
+      
+      femv->endFill();
+      femv->doOwnedToOwnedPlusShared(Tpetra::ADD);
+      femv->doOwnedPlusSharedToOwned(Tpetra::ADD);
+
+      //detect any change in status due to communication
+      for(size_t i = 0; i < nLocalOwned; i++){
+        if(femvData[i].getGroundingStatus() != old_statuses[i]){
+          if(femvData[i].is_art){
+            artQueue.push(i);
+          } else {
+            regQueue.push(i);
+          }
+        }
+      }
+      int local_done = regQueue.empty() && artQueue.empty();
+      //this call makes sure that if any inter-processor communication changed labels
+      //we catch the changes and keep propagating them.
+      Teuchos::reduceAll<int,int>(*comm,Teuchos::REDUCE_MIN,1, &local_done,&done);
+    }
+    
+    
+  }
+  
+  //function that exchanges labels between two nodes
+  //curr_node gives its labels to neighbor.
+  void giveLabels(IcePropVtxLabel<lno_t,gno_t>& curr_node, IcePropVtxLabel<lno_t,gno_t>& neighbor){
+    IcePropGrounding_Status curr_gs = curr_node.getGroundingStatus();
+    IcePropGrounding_Status nbor_gs = neighbor.getGroundingStatus();
+    int curr_node_gid = mapWithCopies->getGlobalElement(curr_node.id);
+    //if the neighbor is full, we don't need to pass labels
+    if(nbor_gs == ICEPROPGS_FULL) return;
+    //if the current nod is empty (shouldn't happen) we can't pass any labels
+    if(curr_gs == ICEPROPGS_NONE) return;
+    //if the current node is full (and not an articulation point), pass both labels on
+    if(curr_gs == ICEPROPGS_FULL && !curr_node.is_art){
+      neighbor.first_label = curr_node.first_label;
```

Comment:
Would using the overloaded += operator here provide greater readability?

Reply:
The += operator considers fewer cases than the give_labels function. I'm not sure using the += operator would provide greater readability, especially since I've had issues before with the queues and cyclic references with the Vertex Label class. Making the += operator have the same functionality as give_labels may lead to more complexity that is distributed over a few files. i think having the propagation rules and functions all in one place makes the code easier to understand.